Direct access to device key

offirko opened the pull request:
### Description

Low code size tool that provides direct access to ROT Device Key
(to be used by BootLoader).

Pending on #9237

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

offirko commented:
@cmonr , @0xc0170 , @ARMmbed/mbed-os-storage - Please review

ciarmcom commented:
@offirko, thank you for your changes.
@ARMmbed/mbed-os-storage @ARMmbed/mbed-os-maintainers please review.

0xc0170 commented on features/storage/kvstore/tdbstore/DirectAccessDevicekey.h:
```diff
@@ -0,0 +1,26 @@
+#include "FlashIAP.h"
+#include <stdio.h>
+
+int direct_access_to_devicekey(uint32_t tdb_start_offset, uint32_t tdb_end_offset, void *data_buf,
```
is this missing doxygen ?

offirko replied:
sorry, I added doxygen info - and squashed the log

0xc0170 replied:
Thanks, looks better. But still can be improved: "inital version" - what initial version is it? what does it bring? or `successfully compiled` ?

offirko replied:
@0xc0170 - I'm not sure I follow.. git is source control. Part of its use is to track the different development stages (e.g. enabling quick recovery milestones..) . I agree that sometimes there are too many unimportant stages that can be squashed .. but in general I don't think we should rewrite git-log history prior to delivery or make it look better)..

0xc0170 replied:
I partly agree but my point was that this feels like development branch rather than ready for the release. Don't need to squash but rather add details important for users.
Two commits msg have zero value without studying changes and this is functionality change, we should introduce it - why are we bringing this functionality change.

offirko replied:
Thanks, I think I now understand what you meant - I'll add the required information.

0xc0170 reviewed (changes requested):
Also git log might be cleaned up (reduced to few c ommits for 3 new files).

cmonr commented:
Fixes https://github.com/ARMmbed/mbed-os/issues/9196

Commit:
Adding low code size tool for direct access to Devicekey.
This enables application with code size restrictions to access
devicekey directly based on address in internal flash without kvconfig overhead

Commit:
Updated test to consider tdb_internal non default configuration as well

offirko commented:
Pending on #9237

bulislaw commented:
Does it need change to design docs and storage documentation? It's hard to see how it fits in the architecture.

dannybenor commented:
This code will not be documented. Is for the use of the PDMC bootloader, to reduce the size and avoid linking the full TDBStore. In 2 versions, the bootloader will completely change to a PSA oriented, and this code will not be relevant.

cmonr commented:
> https://github.com/ARMmbed/mbed-os/pull/9215#issuecomment-451237780

Fyi @ARMmbed/mbed-docs @ARMmbed/mbed-os-maintainers 
Imo, this sounds reasonable.

0xc0170 commented:
Version based on prceeding PR

offirko commented:
@davidsaada - can you please review - Thanks

davidsaada commented on features/storage/kvstore/tdbstore/DirectAccessDevicekey.cpp:
```diff
@@ -0,0 +1,284 @@
+exit_point:
+    if (true == is_flash_init) {
+        flash.deinit();
+        is_flash_init = false;
```
Can remove this line.

davidsaada commented on features/storage/kvstore/tdbstore/DirectAccessDevicekey.cpp:
```diff
@@ -0,0 +1,284 @@
+        *out_tdb_start_offset =  MBED_CONF_STORAGE_TDB_EXTERNAL_INTERNAL_BASE_ADDRESS;
+        tdb_size = MBED_CONF_STORAGE_TDB_EXTERNAL_RBP_INTERNAL_SIZE;
+    } else {
+        return MBED_ERROR_UNSUPPORTED;
```
Should deinit flash here and on next failures.

offirko replied:
my bad, some of the ifs are later addition to the code. I will move flash init below them.

davidsaada commented on features/storage/kvstore/tdbstore/DirectAccessDevicekey.cpp:
```diff
@@ -0,0 +1,284 @@
+static int reserved_data_get(FlashIAP *flash, tdbstore_area_data_t *area_params, void *reserved_data_buf,
+                             size_t reserved_data_buf_size, size_t *actual_data_size_ptr)
+{
+    int status = MBED_SUCCESS;;
```
redundant semicolon.

davidsaada reviewed (changes requested):
Looks good, except for a few minor issues.

Commit:
Added utility function to help locate expected internal flash tdbstore position.
(Supporting FILESYSTEM and TDB_EXTERNAL configuration only)

0xc0170 commented:
Ready for CI as soon as reviewers approve (@davidsaada )

davidsaada reviewed (approved):
Looks good now.

0xc0170 commented:
CI started

Commit:
The direct_access_devicekey test should not run if there is no component flashiap block device in the target

yossi2le commented:
I have changed the test main file to skip compiling if there is no component flashiap set for the target. I hope this will solve all the build failure. 
Can you please run the CI again?
Thanks

offirko commented:
@yossi2le - thanks for the fix!
@0xc0170 - travis failure doesnt seem to be related - can u pls rerun.

NirSonnenschein commented:
Hi @offirko,
The Travis failures are caused by an issue fixed on master a few hours ago, and this change should be propagated here via rebase. We can start CI only after travis is fixed, could you please rebase this PR?
Sorry for the trouble

0xc0170 commented:
CI started
